### PR TITLE
Changing jungle trees a bit

### LINKED ISSUE #1349: Add variation to Jungle Trees
Every jungle tree has the same model... Add more "floors" of leaves!

### PR DESCRIPTION
**Pull request by 12xx12**
So I had a look at old issues and someone asked for new jungle tree models 

Fixes #1349

I took a look around in vanilla and the implemented behaviour is quite close to vanilla. I changed some numbers to get closer and moved and renamed a variable to make more clear what it does.

### CODE REVIEW

**Comment by tigerw** [comment]
Before and after screenshot? :)

**Comment by 12xx12** [comment]
Old


New


Vanilla

**Comment by 12xx12** [comment]
After revisiting I think the threshold for branches could be even lower

File: src/Generating/Trees.cpp [hunk 1]
```diff
@@ -58,7 +59,8 @@ static const Vector3d & pickBranchDirection(const cNoise a_Noise, Vector3i a_Blo
 		}
 	};
 
-	size_t index = static_cast<size_t>(a_Noise.IntNoise3DInt(a_BlockPos.x, a_BlockPos.y + a_Seq, a_BlockPos.z)) % directions.size();
+	// Old: size_t index = static_cast<size_t>(a_Noise.IntNoise3DInt(a_BlockPos.x, a_BlockPos.y + a_Seq, a_BlockPos.z)) % directions.size();
```
Comment (12xx12) [inline src/Generating/Trees.cpp:62]: This resulted in many trees having the branches in the same/close direction

**Comment by 12xx12** [comment]
New (V2)

**Comment by 12xx12** [comment]
Another thing I noticed that the branches on vanilla only grow 90 Degrees from the trunk surface not diagonal

File: src/Generating/Trees.cpp (hunk 1 above)
Comment (12xx12) [reply src/Generating/Trees.cpp:62]: I just noticed that the function descriptions is not accurate anymore and that the passed values are not used at all

File: src/Generating/Trees.cpp [hunk 2]
```diff
@@ -58,7 +59,8 @@ static const Vector3d & pickBranchDirection(const cNoise a_Noise, Vector3i a_Blo
...
+	size_t index = static_cast<size_t>(GetRandomProvider().RandInt(directions.size() - 1));
```
Comment (peterbell10) [inline src/Generating/Trees.cpp:63]: You can't use randomness is the terrain generator, if you do then world seeds won't reproduce.

Comment (12xx12) [reply src/Generating/Trees.cpp:63]: Fair Point. Back to the drawing board

Comment (12xx12) [reply src/Generating/Trees.cpp]: changed with new comit

File: src/Generating/Trees.cpp [hunk 3]
```diff
@@ -1299,26 +1300,41 @@ void GetLargeJungleTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, s
 		a_OtherBlocks.push_back(sSetBlock(a_BlockPos.addedY(-i).addedXZ(1, 1), E_BLOCK_DIRT, E_META_DIRT_NORMAL));
 	}
 
-	int numBranches = (a_Noise.IntNoise2DInt(a_BlockPos.x * a_Seq, a_BlockPos.z * a_Seq) / 11) % 3 + 1;
-	int branchStartHeight = 8 + Height % 11;
-	int branchInterval = (Height - branchStartHeight) / numBranches;
-	for (int i = branchStartHeight; i < (Height - 6); i += branchInterval)
+	int NumBranches = std::max(
+		(a_Noise.IntNoise2DInt(a_BlockPos.x * a_Seq, a_BlockPos.z * a_Seq) / 10) % 4,  // The Original Calculation
+		FloorC(Height / (Height / 2.0f))                                               // Just to assure that no massive trees spawn with just one branch
```
Comment (peterbell10) [inline src/Generating/Trees.cpp:1305]: Isn't this just 2?

File: src/Generating/Trees.cpp [hunk 4]
```diff
@@ -1299,26 +1300,41 @@ void GetLargeJungleTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, s
...
 		Vector3d BranchDirection = pickBranchDirection(a_Noise, a_BlockPos.addedY(i * a_Seq), a_Seq) / 3;
-
 		int BranchLength = 2 + a_Noise.IntNoise3DInt(a_BlockPos * a_Seq) % 2;
 		Vector3i BranchEndPosition = GetTreeBranch(E_BLOCK_LOG, E_META_LOG_JUNGLE, a_BlockPos.addedY(i), BranchLength, BranchStartDirection, BranchDirection, a_LogBlocks).Floor();
-		PushCoordBlocks(BranchEndPosition.x, BranchEndPosition.y, BranchEndPosition.z, a_OtherBlocks, BigO2, ARRAYCOUNT(BigO2), E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE);
-		PushCoordBlocks(BranchEndPosition.x, BranchEndPosition.y + 1, BranchEndPosition.z, a_OtherBlocks, BigO1, ARRAYCOUNT(BigO1), E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE);
-		a_OtherBlocks.push_back(sSetBlock(BranchEndPosition.x, BranchEndPosition.y + 1, BranchEndPosition.z, E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE));
+
+		// There's a chance that there is a third leaf level on a branch
+		if (a_Noise.IntNoise3DInt(a_BlockPos.x * a_Seq, a_BlockPos.y * a_Seq, a_BlockPos.z * a_Seq) % 3 > 2)  // A quarter chance
```
Comment (peterbell10) [inline src/Generating/Trees.cpp:1318]: `a % 3` can never be `> 2`. I think you mean `a % 4 > 2` though you might as well say `== 0` which is probably easier to think about.

File: src/Generating/Trees.cpp [hunk 5]
```diff
@@ -58,7 +59,7 @@ static const Vector3d & pickBranchDirection(const cNoise a_Noise, Vector3i a_Blo
 		}
 	};
 
-	size_t index = static_cast<size_t>(a_Noise.IntNoise3DInt(a_BlockPos.x, a_BlockPos.y + a_Seq, a_BlockPos.z)) % directions.size();
+	size_t index = static_cast<size_t>(a_Noise.IntNoise3DInt(a_BlockPos.x * a_Seq, a_BlockPos.y * a_Seq, a_BlockPos.z * a_Seq)) % directions.size();
```
Comment (peterbell10) [inline src/Generating/Trees.cpp:62]: Are you concerned about overflow here? How large does `a_Seq` get?

File: src/Generating/Trees.cpp (hunk 4 above)
Comment (12xx12) [reply src/Generating/Trees.cpp:1318]: yes it is easier - changed

File: src/Generating/Trees.cpp (hunk 5 above)
Comment (12xx12) [reply src/Generating/Trees.cpp:62]: No i'm not

i am now takinng  ```a_Noise.IntNoise3DInt(a_BlockPos * a_Seq)``` from line 1314 now - this is old code which worked

File: src/Generating/Trees.cpp (hunk 3 above)
Comment (12xx12) [reply src/Generating/Trees.cpp:1305]: yes

i changed it that for every 10 blocks of height at least one branch is spawned (maximum is at 29) 

0-9 minimum is 0
10-19 minimum is 1
20-29 minimum is 2

File: src/Generating/Trees.cpp [hunk 6]
```diff
@@ -1299,26 +1300,41 @@ void GetLargeJungleTreeImage(Vector3i a_BlockPos, cNoise & a_Noise, int a_Seq, s
...
 		Vector3d BranchDirection = pickBranchDirection(a_Noise, a_BlockPos.addedY(i * a_Seq), a_Seq) / 3;
-
 		int BranchLength = 2 + a_Noise.IntNoise3DInt(a_BlockPos * a_Seq) % 2;
 		Vector3i BranchEndPosition = GetTreeBranch(E_BLOCK_LOG, E_META_LOG_JUNGLE, a_BlockPos.addedY(i), BranchLength, BranchStartDirection, BranchDirection, a_LogBlocks).Floor();
-		PushCoordBlocks(BranchEndPosition.x, BranchEndPosition.y, BranchEndPosition.z, a_OtherBlocks, BigO2, ARRAYCOUNT(BigO2), E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE);
-		PushCoordBlocks(BranchEndPosition.x, BranchEndPosition.y + 1, BranchEndPosition.z, a_OtherBlocks, BigO1, ARRAYCOUNT(BigO1), E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE);
-		a_OtherBlocks.push_back(sSetBlock(BranchEndPosition.x, BranchEndPosition.y + 1, BranchEndPosition.z, E_BLOCK_LEAVES, E_META_LEAVES_JUNGLE));
+
+		// There's a chance that there is a third leaf level on a branch
+		if (a_Noise.IntNoise3DInt(a_BlockPos.x * a_Seq, a_BlockPos.y * a_Seq, a_BlockPos.z * a_Seq) == 0)  // A quarter chance
```
Comment (peterbell10) [inline src/Generating/Trees.cpp:1318]:
```suggestion
		if ((a_Noise.IntNoise3DInt(a_BlockPos.x * a_Seq, a_BlockPos.y * a_Seq, a_BlockPos.z * a_Seq) % 4) == 0)  // A quarter chance
```

**Comment by bearbin** [comment]
@peterbell10 Do you have any more concerns here or is this good to merge now?

**Review by peterbell10 — APPROVED** [review]

**Pull request merged**